content/blog/2020-12-20-loop-flatten/index.md

```diff
@@ -0,0 +1,199 @@
+
+
+## Implementation
+The idea is quite straightforward but the implementation is not that easy. To implement this pass, we should perform the following modifications.
+
+1. Update the loop bound and increment <br/>
+As the first step, we should update the loop bound for the outer loop. Suppose the loop bound for the inner loop and outer loop are m and n with the same increment, we can change the loop bound for the outer loop to be `m*n` and keep the same increment.
+
+2. Extract the statements inside the inner loop to the outer loop<br/>
+In this step, we extract all the statements inside the inner loop and put them into the outer loop. At the same time, we should detect the usage of the inner loop variable and then remove it with the correct one after flattening the loop. For example, suppose we have a piece of code like this inside the inner loop: `i*n+j`, whereas `i` and `j` are the loop variables, `n` is the outer loop bound, and the increment for both inner loop and outer loop is 1. To correctly flatten this loop, we should replace the statement `i*n+j` with `i`. 
+
+3. Delete the inner loop<br/>
+After replacing the statement from the inner loop to the outer loop, we can safely remove the inner loop by deleting the corresponding basic blocks.
+
+Finally, we should check the equivalence of the code, i.e., the flattened loop should perform the same execution as the original nested loop. This can be verified by testing the pass with benchmarks. 
+
+During the implementation, I got stuck in the second step, i.e., extract the statements inside the inner loop to the outer loop. The generated code performs weird when extracts the loop variant expressions from the inner loop to the outer loop. 
+
+Thus, to simplify the problem, I decide to change the inner loop bound instead of the outer loop bound. This is quite easy because we only need the first step for “flattening” the loop. But for the pass I build, it only works for the “perfect” nested loop, i.e., the nest loop that only contains statements inside the inner loop. 
+```javascript
+for (int i = 0; i < n; i++) { // unflattened “perfect” nested loop
+    // no statement here
+    for (int j = 0; j < m; j++) {
+      Stmt(f(i, j))
+```
+
+My naïve pass is available in [this link](https://github.com/hj424/bril/tree/master/tasks/project/llvm-pass-skeleton).
+
+## Evaluation
+1. Correctness <br/>
+As for the correctness of the pass, we should focus on two points: the functionality of the optimized code and the correctness of the unrolled loop. These two points are being checked by running the simple benchmarks with the clang profiling tools. For more details, please check out [this link](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html).

```

Comment:
"the functionality of the optimized code and the correctness of the unrolled loop" sound like the same thing? Maybe this is supposed to be the correctness and the performance?